replay/models/nn/sequential/sasrec_with_llm/lightning.py

```diff
@@ -0,0 +1,361 @@
+                         use_upscale=use_upscale,
+                         weight_scale=weight_scale,
+                         multi_profile=multi_profile,
+                         multi_profile_aggr_scheme=multi_profile_aggr_scheme)
+        self.scale_guide_loss = scale_guide_loss
+        self.alpha = alpha
+        self.profile_distil_epochs = profile_distil_epochs
+
+    def _compute_loss(self, batch: SasRecLLMTrainingBatch) -> torch.Tensor:
+        # loss_func # nn mse
+
+        if self._loss_type == "BCE":
+            loss_func = self._compute_loss_bce if self._loss_sample_count is None else self._compute_loss_bce_sampled
+        elif self._loss_type == "CE":
+            loss_func = self._compute_loss_ce if self._loss_sample_count is None else self._compute_loss_ce_sampled
+        else:
+            msg = f"Not supported loss type: {self._loss_type}"
+            raise ValueError(msg)
+
+        # returns hidden states additionally
+        loss_model, hidden_state = loss_func(
+            batch.features,
+            batch.labels,
+            batch.padding_mask,
+            batch.labels_padding_mask,
+        )
+
+        # reconstruction loss
+        loss_guide = self._compute_loss_reconstruction(
+            batch.user_profile_embeddings_batch,
+            hidden_state,
+            batch.null_profile_binary_mask_batch,
+            torch.nn.MSELoss(),
+        )
+        return self._merge_losses(loss_model, loss_guide)
+
+    def _merge_losses(self, loss_model: torch.Tensor, loss_guide: torch.Tensor) -> torch.Tensor:
+        epoch = self.trainer.current_epoch
+        if self.scale_guide_loss:
+            loss_model_val = loss_model.item()
+            loss_guide_val = loss_guide.item()
+            eps = 1e-8
+
+            scale_for_guide = loss_model_val / (loss_guide_val + eps)
+            scaled_loss_guide = loss_guide * scale_for_guide
+
+            if epoch < self.profile_distil_epochs:
+                loss = self.alpha * scaled_loss_guide + (1 - self.alpha) * loss_model
+            else:
+                loss = loss_model
+        else:
+            # Если self.scale_guide_loss=False, логика остаётся исходной
+            if epoch < self.profile_distil_epochs:
+                loss = self.alpha * loss_guide + (1 - self.alpha) * loss_model
+            else:
+                loss = loss_model
+        return loss
+
+    def _compute_loss_reconstruction(
+        self,
+        user_profile_emb: torch.Tensor,
+        hidden_for_reconstruction: torch.Tensor,
+        null_profile_binary_mask_batch: torch.BoolTensor,
+        criterion_reconstruct_fn
+    ) -> torch.Tensor:
+        user_profile_emb_transformed = self._model.aggregate_profile(user_profile_emb)
+        if self._model.use_upscale:
+            hidden_for_reconstruction = self._model.hidden_layer_transform(hidden_for_reconstruction)
+
+        null_profile_binary_mask_batch = null_profile_binary_mask_batch.flatten()
```

Comment:
Seems like this line equalizes null profiles with corresponding hidden_for_reconstruction. 
It seems more convinient than assigning zero embedding, but why not skip null profiles in loss?